db/src/tx.rs

```diff
@@ -263,8 +267,24 @@ impl<'conn, 'a> Tx<'conn, 'a> {
         // And evolve them forward.
         while generation.can_evolve() {
             // Evolve further.
-            let temp_id_map = self.resolve_temp_id_avs(&generation.temp_id_avs()[..])?;
+            let temp_id_map: TempIdMap = self.resolve_temp_id_avs(&generation.temp_id_avs()[..])?;
             generation = generation.evolve_one_step(&temp_id_map);
+
+            // Report each tempid that resolves via upsert.
+            for (tempid, &entid) in &temp_id_map {
+                // Every tempid should be resolved at most once.  Prima facie, we might expect a
+                // tempid to be resolved in two different generations.  However, that is not so: the
+                // keys of temp_id_map are unique between generations.Suppose that id->e and id->e*
+                // are two such mappings, resolved on subsequent evolutionary steps, and that `id`
+                // is a key in the intersection of the two key sets. This can't happen: if `id` maps
+                // to `e` via id->e, all instances of `id` have been evolved forward (replaced with
+                // `e`) before we try to resolve the next set of `UpsertsE`.  That is, we'll never
+                // successfully upsert the same tempid in more than one generation step.  (We might
+                // upsert the same tempid to multiple entids via distinct `[a v]` pairs in a single
+                // generation step; in this case, the transaction will fail.)
+                assert!(!tempids.contains_key(&**tempid));
+                tempids.insert((**tempid).clone(), entid);
```

Comment:
`insert` returns `None` if there was no existing value for that key. You can use that instead of the separate `contains_key` check:

```rust
let previous = tempids.insert(…);
assert!(previous.is_none());
```